modules/assim.sequential/R/SDA_NA_downscale.R

```diff
@@ -0,0 +1,459 @@
+  # load maps.
+  all.rast <- terra::rast(rast.dir)
+  # parallel loop.
+  layer.names <- names(all.rast)
+  # register parallel nodes.
+  if (cores > length(layer.names)) {
+    cores <- length(layer.names)
+  }
+  cl <- parallel::makeCluster(as.numeric(cores))
+  doSNOW::registerDoSNOW(cl)
+  #progress bar.
+  pb <- utils::txtProgressBar(min=1, max=length(layer.names), style=3)
+  progress <- function(n) utils::setTxtProgressBar(pb, n)
+  opts <- list(progress=progress)
+  # foreach loop.
+  vecs <- foreach::foreach(r = seq_along(layer.names), 
+                           .packages=c("Kendall", "terra"),
+                           .options.snow=opts) %dopar% {
+                             all.rast <- terra::rast(rast.dir)
+                             temp.vec <- matrix(all.rast[[r]], byrow = T)
+                             na.inds <- which(is.na(temp.vec))
+                             # if it's LC layer.
+                             if ("LC" == names(all.rast)[r]) {
+                               non.veg.inds <- which(! temp.vec %in% 1:8)
+                               na.inds <- unique(c(na.inds, non.veg.inds))
+                             }
+                             return(list(vec = temp.vec,
+                                         na.inds = na.inds))
+                           }
+  # stop parallel.
+  parallel::stopCluster(cl)
+  foreach::registerDoSEQ()
+  gc()
+  # grab uniqued NA index.
+  na.inds <- vecs %>% purrr::map("na.inds") %>% unlist %>% unique
+  # remove NA from each covariate.
+  cov.vecs <- vecs %>% purrr::map(function(v){
+    return(v$vec[-na.inds])
+  }) %>% dplyr::bind_cols() %>% `colnames<-`(layer.names) %>% as.data.frame()
+  non.na.inds <- seq_along(matrix(all.rast[[1]]))[-na.inds]
+  return(list(df = cov.vecs, non.na.inds = non.na.inds))
+}
+
+#' @description
+#' This function helps to create the training dataset of specific variable type and locations for downscaling.
+#' TODO: There will be a ratio argument between training and testing samples to testify the ML regression accuracy.
+#' @title prepare.train.dat
+#' 
+#' @param settings character: physical path that points to the pecan settings XML file.
+#' @param analysis numeric: data frame (rows: ensemble member; columns: site*state_variables) of updated ensemble analysis results from the `sda_enkf` function.
+#' @param covariates.dir character: path to the exported covariates GeoTIFF file.
+#' @param variable character: name of state variable. It should match up with the column names of the analysis data frame. 
+#'
+#' @return matrix within which the first sets of columns contain values of state variables for each ensemble mebers of every site, and the rest columns contain the corresponding covariates.
+#' 
+#' @author Dongchen Zhang
+prepare.train.dat <- function(settings, analysis, covariates.dir, variable) {
+  # convert settings into geospatial points.
+  pts <- pecan.settings.2.pts(settings)
```

Comment:
If the settings object is only used to get the points, why not just have the points be the function argument? That gives you more flexibility to have pulled the points from the csv or shapefile

Reply:
Fixed.